src/native/minipal/utf8.c

```diff
@@ -1,2150 +1,669 @@
-                (chc & 0xC0) != 0x80)
-            {
-                goto BadLongCode;
-            }
-
-            chc &= 0x3F;
-
-            // start a new long code
-            if ((ch & 0x20) != 0)
-            {
-                // fold the first two bytes together
-                chc |= (ch & 0x0F) << 6;
-
-                if ((ch & 0x10) != 0)
-                {
-                    // 4 byte encoding - surrogate
-                    ch = *pSrc;
-                    if (
-                        // check that bit 4 is zero, the non-shortest form of surrogate
-                        // and the valid surrogate range 0x000000 - 0x10FFFF at the same time
-                        !InRange(chc >> 4, 0x01, 0x10) ||
-                        // we are expecting to see trailing bytes like 10vvvvvv
-                        (ch & 0xC0) != 0x80)
-                    {
-                        goto BadLongCode;
-                    }
-
-                    chc = (chc << 6) | (ch & 0x3F);
-
-                    ch = *(pSrc + 1);
-                    // we are expecting to see trailing bytes like 10vvvvvv
-                    if ((ch & 0xC0) != 0x80)
-                    {
-                        goto BadLongCode;
-                    }
-                    pSrc += 2;
-
-                    // extra byte
-                    charCount--;
-                }
-                else
-                {
-                    // 3 byte encoding
-                    ch = *pSrc;
-                    if (
-                        // check for non-shortest form of 3 byte seq
-                        (chc & (0x1F << 5)) == 0 ||
-                        // Can't have surrogates here.
-                        (chc & (0xF800 >> 6)) == (0xD800 >> 6) ||
-                        // we are expecting to see trailing bytes like 10vvvvvv
-                        (ch & 0xC0) != 0x80)
-                    {
-                        goto BadLongCode;
-                    }
-                    pSrc++;
-
-                    // extra byte
-                    charCount--;
-                }
-            }
-            else
-            {
-                // 2 byte encoding
-
-                // check for non-shortest form
-                if ((ch & 0x1E) == 0) goto BadLongCode;
-            }
-
-            // extra byte
-            charCount--;
+        #if BIGENDIAN
+        if (!treatAsLE)
+        {
+            currentChar = BYTESWAP16(currentChar);
         }
+        #endif
 
-        // no pending bits at this point
-        ch = 0;
-        continue;
-
-    BadLongCode:
-        pSrc -= 2;
-        ch = 0;
-        continue;
+        destination[destinationIndex++] = (CHAR16_T)currentChar;
     }
 
-    // May have a problem if we have to flush
-    if (ch != 0)
+    if (sourceIndex < sourceLength && (destinationIndex > destinationLength || (destinationIndex == destinationLength && sourceIndex + 1 < sourceLength)))
```

Comment:
@lambdageek, just an FYI, I had to change this condition from

`    if (sourceIndex < sourceLength && (destinationIndex >= destinationLength)`

for mono because this case was failing:

https://github.com/dotnet/runtime/blob/2bfd1b44200de244a66b7f4a7604be1f2d5585bc/src/libraries/System.Runtime/tests/System.Runtime.Tests/System/AttributeTests.cs#L169

IMHO, the reflection stack of mono should use the same expectation from giconv about the expected lengths as coreclr's reflection stack makes from coreclr PAL's unicode.cpp, so this standalone component doesn't need to handle more edge cases than necessary.

Reply:
/cc @fanyang-mono

Reply:
The test that @am11 mentioned was testing the custom attribute value string "\uDFFF". For mono, it went through `minipal_convert_utf8_to_utf16` to decode this string "\xed\xbf\xbf". It seems to me that CoreCLR went through a different mechanism to get the value of the custom attribute value.

Reply:
@fanyang-mono, coreclr ultimately uses the same implementation (couple of wrappers later), but in reflection stack (metasig.h etc.), there it compensates for those off-by-one, zero-length kind of scenarios to match the expectation.

Reply:
@am11 Do you mind pointing me to the CoreCLR source code where they handles off-by-one, zero-length kind of scenarios?

Reply:
@fanyang-mono, good question. 😅  It is a bit tricky to discern the exact place without actually attaching a conditional debugger (or some tricky way), but I was debugging the similar tests for coreclr last year https://github.com/dotnet/runtime/pull/85558#discussion_r1234870799 which points to https://github.com/dotnet/runtime/blob/d8c59a418e0857d2bdf71eea064cd6e55a093092/src/coreclr/md/compiler/regmeta_import.cpp#L705
One thing I do remember; when all PAL tests were passing and 99% of managed tests were passingm those edge cases covered in reflection tests were still failing, which is why I had to add a similar special condition in unicode.cpp -> utf8.c conversion to balance varied expectations of mono and coreclr + reflection simultaneously.

Reply:
It is probably fine having this a little bit longer check, because to make mono align with CoreCLR on this is non-trivial work.